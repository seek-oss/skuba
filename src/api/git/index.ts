export { commit } from './commit';
export { commitAllChanges } from './commitAllChanges';
export { currentBranch } from './currentBranch';
export { findRoot } from './findRoot';
export type { ChangedFile } from './getChangedFiles';
export { getChangedFiles } from './getChangedFiles';
export { getHeadCommitId, getHeadCommitMessage } from './log';
export { getOwnerAndRepo } from './remote';
export { push } from './push';
export { fastForwardBranch } from './pull';
export { reset } from './reset';
export { isFileGitIgnored } from './isFileGitIgnored';
