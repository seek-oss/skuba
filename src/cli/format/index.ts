import chalk from 'chalk';

import { hasDebugFlag } from '../../utils/args.js';
import { createLogger, log } from '../../utils/logging.js';
import { runESLint } from '../adapter/eslint.js';
import { runPrettier } from '../adapter/prettier.js';
import { internalLint } from '../lint/internal.js';

export const format = async (
  args = process.argv.slice(2),
  overrideConfigFile?: string,
): Promise<void> => {
  const debug = hasDebugFlag(args);

  log.plain(chalk.blueBright('skuba lints'));

  const internal = await internalLint('format', {
    debug,
    additionalFlags: args.includes('--force-apply-all-patches')
      ? ['--force-apply-all-patches']
      : [],
    serial: true,
  });

  const logger = createLogger({ debug });

  log.newline();
  log.plain(chalk.magenta('ESLint'));

  const eslint = await runESLint('format', logger, overrideConfigFile);

  log.newline();
  log.plain(chalk.cyan('Prettier'));

  const prettier = await runPrettier('format', logger);

  if (eslint.ok && prettier.ok && internal.ok) {
    return;
  }

  const tools = [
    ...(eslint.ok ? [] : ['ESLint']),
    ...(prettier.ok ? [] : ['Prettier']),
    ...(internal.ok ? [] : ['skuba']),
  ];

  log.newline();
  log.err(tools.join(', '), 'found issues that require triage.');

  process.exitCode = 1;
};
