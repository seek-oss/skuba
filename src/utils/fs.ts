import fs from 'fs-extra';

import { isErrorWithCode } from './error.js';

export const pathExists = async (filePath: string): Promise<boolean> => {
  try {
    await fs.promises.access(filePath);

    return true; // Path exists and is accessible
  } catch (error: unknown) {
    if (isErrorWithCode(error, 'ENOENT')) {
      return false; // Path does not exist
    }

    throw error; // Other errors (include permission issues)
  }
};
