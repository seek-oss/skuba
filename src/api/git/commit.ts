import fs from 'fs-extra';

export interface Identity {
  email?: string;
  name?: string;
}

interface CommitParameters {
  author?: Identity;
  committer?: Identity;
  dir: string;
  message: string;
}

/**
 * Writes a commit to the local Git repository.
 */
export const commit = async ({
  author = { name: 'skuba' },
  committer = { name: 'skuba' },
  dir,
  message,
}: CommitParameters): Promise<string> => {
  const git = await import('isomorphic-git');
  return git.commit({
    author,
    committer,
    dir,
    fs,
    message,
  });
};
