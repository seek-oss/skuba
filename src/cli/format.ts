import chalk from 'chalk';

import { hasDebugFlag } from '../utils/args';
import { createLogger, log } from '../utils/logging';

import { runESLint } from './adapter/eslint';
import { runPrettier } from './adapter/prettier';
import { tryAddEmptyExports } from './configure/addEmptyExports';
import { tryPatchRenovateConfig } from './configure/patchRenovateConfig';
import { tryRefreshIgnoreFiles } from './configure/refreshIgnoreFiles';

export const format = async (args = process.argv.slice(2)): Promise<void> => {
  await Promise.all([
    tryAddEmptyExports(),
    tryPatchRenovateConfig(),
    tryRefreshIgnoreFiles(),
  ]);

  const debug = hasDebugFlag(args);
  const logger = createLogger(debug);

  log.plain(chalk.magenta('ESLint'));

  const eslint = await runESLint('format', logger);

  log.newline();
  log.plain(chalk.cyan('Prettier'));

  const prettier = await runPrettier('format', logger);

  if (eslint.ok && prettier.ok) {
    return;
  }

  const tools = [
    ...(eslint.ok ? [] : ['ESLint']),
    ...(prettier.ok ? [] : ['Prettier']),
  ];

  log.newline();
  log.err(tools.join(', '), 'found issues that require triage.');

  process.exitCode = 1;
};
