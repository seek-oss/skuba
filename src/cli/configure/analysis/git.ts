import fs from 'fs-extra';
import git from 'isomorphic-git';

import { crawlDirectory } from '../../../utils/dir';
import { log } from '../../../utils/logging';

export const auditWorkingTree = async (dir: string) => {
  const filepaths = await crawlDirectory(dir);

  const statuses = await Promise.all(
    filepaths.map((filepath) => git.status({ dir, fs, filepath })),
  );

  if (
    statuses.some(
      (status) =>
        status !== 'absent' && status !== 'ignored' && status !== 'unmodified',
    )
  ) {
    log.newline();
    log.warn('You have dirty/untracked files that may be overwritten.');
  }
};
