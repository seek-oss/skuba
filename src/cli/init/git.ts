import path from 'path';

import fs from 'fs-extra';
import git from 'isomorphic-git';
import simpleGit from 'simple-git';

import * as Git from '../../api/git';
import { log } from '../../utils/logging';

interface GitHubProject {
  orgName: string;
  repoName: string;
  defaultBranch: string;
}

export const initialiseRepo = async (
  dir: string,
  { orgName, repoName, defaultBranch }: GitHubProject,
) => {
  await git.init({
    defaultBranch,
    dir,
    fs,
  });

  await Git.commit({
    dir,
    message: 'Initial commit',
  });

  await git.addRemote({
    dir,
    fs,
    remote: 'origin',
    url: `git@github.com:${orgName}/${repoName}.git`,
  });
};

export const downloadGitHubTemplate = async (
  gitHubPath: string,
  destinationDir: string,
) => {
  log.newline();
  log.plain('Downloading', log.bold(gitHubPath), 'from GitHub...');

  await simpleGit().clone(`git@github.com:${gitHubPath}.git`, destinationDir, [
    '--depth=1',
    '--quiet',
  ]);

  await fs.promises.rm(path.join(destinationDir, '.git'), {
    force: true,
    recursive: true,
  });

  log.newline();
  log.warn(
    'You may need to run',
    log.bold('yarn skuba configure'),
    'once this is done.',
  );
};
