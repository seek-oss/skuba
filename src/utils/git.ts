import fs from 'fs-extra';
import git from 'isomorphic-git';
import http from 'isomorphic-git/http/node';

export const gitCommit = async ({
  dir,
  message,
}: {
  dir: string;
  message: string;
}) =>
  git.commit({
    author: { name: 'skuba' },
    committer: { name: 'skuba' },
    dir,
    fs,
    message,
  });

interface GitPushProps {
  auth: { type: 'gitHubApp'; token: string };
  branch: string;
  commitOid: string;
  dir: string;
}

export const gitPush = async ({
  auth,
  branch,
  commitOid,
  dir,
}: GitPushProps) => {
  const { owner, repo } = await getOwnerRepo(dir);

  const url = `https://github.com/${encodeURIComponent(
    owner,
  )}/${encodeURIComponent(repo)}`;

  return git.push({
    onAuth: () => ({
      username: 'x-access-token',
      password: auth.token,
    }),
    http,
    dir,
    fs,
    ref: commitOid,
    remoteRef: branch,
    url,
  });
};

export const getHeadSha = async (dir: string): Promise<string> => {
  const [commit] = await git.log({ depth: 1, dir, fs });

  return commit.oid;
};

/**
 * Matches the owner and repository names in a GitHub repository URL.
 *
 * For example, given the following input strings:
 *
 * ```console
 * git@github.com:seek-oss/skuba.git
 * https://github.com/seek-oss/skuba.git
 * ```
 *
 * This pattern will produce the following matches:
 *
 * 1. seek-oss
 * 2. skuba
 */
const ownerRepoRegex = /github.com(?::|\/)(.+)\/(.+).git$/;

export const getOwnerRepo = async (
  dir: string,
): Promise<{ owner: string; repo: string }> => {
  const remotes = await git.listRemotes({ dir, fs });

  for (const { url } of remotes) {
    const match = ownerRepoRegex.exec(url);

    const owner = match?.[1];
    const repo = match?.[2];

    if (owner && repo) {
      return { owner, repo };
    }
  }

  throw new Error('Could not find a GitHub remote');
};
