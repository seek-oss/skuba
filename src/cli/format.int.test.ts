import crypto from 'crypto';
import path from 'path';

import fs, { copy } from 'fs-extra';
import git from 'isomorphic-git';
import { diff } from 'jest-diff';

import { format } from './format';

jest.setTimeout(15_000);

const stdoutMock = jest.fn();

jest
  .spyOn(console, 'log')
  .mockImplementation((...args) => stdoutMock(`${args.join(' ')}\n`));

const SOURCE_FILES = ['a/a/a.ts', 'b.md', 'c.json', 'd.js'];

const BASE_PATH = path.join(__dirname, '..', '..', 'integration', 'base');

const TEMP_PATH = path.join(__dirname, '..', '..', 'integration', 'format');

const stdout = (randomMatcher: RegExp) =>
  stdoutMock.mock.calls
    .flat(1)
    .join('')
    .replace(/ in [\d\.]+s\./g, ' in <random>s.')
    .replace(randomMatcher, '<random>');

const gitAdd = (dir: string) =>
  Promise.all(
    SOURCE_FILES.map((filepath) =>
      git.add({
        dir,
        filepath,
        fs,
      }),
    ),
  );

const gitModifiedAndUnstaged = async (dir: string) => {
  const matrix = await git.statusMatrix({
    dir,
    filepaths: SOURCE_FILES,
    fs,
  });

  return matrix.flatMap(([filename, head, workdir, stage]) =>
    head === 1 && workdir === 2 && stage === 1 ? filename : [],
  );
};

const prepareTempDirectory = async (baseDir: string, tempDir: string) => {
  await copy(baseDir, tempDir, { recursive: true });

  process.chdir(tempDir);

  await git.init({
    dir: tempDir,
    fs,
  });

  await gitAdd(tempDir);

  const commitOid = await git.commit({
    author: { name: 'skuba' },
    committer: { name: 'skuba' },
    dir: tempDir,
    fs,
    message: 'Initial commit',
  });

  const entries = await git.walk({
    dir: tempDir,
    fs,
    // eslint-disable-next-line new-cap
    trees: [git.TREE({ ref: commitOid })],
    map: async (filename, [file]) => [filename, await file?.oid()],
  });

  return Object.fromEntries(entries);
};

const originalCwd = process.cwd();

beforeEach(jest.clearAllMocks);

afterAll(() => {
  // Restore the original working directory to avoid confusion in other tests.
  process.chdir(originalCwd);

  // Clean up temporary directories to avoid subsequent formatting and linting
  // warnings and to save on disk space. This can be commented out to inspect
  // the output.
  return fs.promises.rm(TEMP_PATH, {
    force: true,
    recursive: true,
  });
});

interface Args {
  args: string[];
  base: string;
  exitCode: number | undefined;
  modified: string[];
}

test.each`
  description     | args           | base           | exitCode     | modified
  ${'fixable'}    | ${[]}          | ${'fixable'}   | ${undefined} | ${['a/a/a.ts', 'b.md', 'c.json', 'd.js']}
  ${'ok'}         | ${[]}          | ${'ok'}        | ${undefined} | ${[]}
  ${'ok --debug'} | ${['--debug']} | ${'ok'}        | ${undefined} | ${[]}
  ${'unfixable'}  | ${[]}          | ${'unfixable'} | ${1}         | ${['a/a/a.ts']}
`('$description', async ({ args, base, exitCode, modified }: Args) => {
  const baseDir = path.join(BASE_PATH, base);

  const tempDir = path.join(
    TEMP_PATH,
    `${base}-${crypto.randomBytes(32).toString('hex')}`,
  );

  const originalFiles = await prepareTempDirectory(baseDir, tempDir);

  await expect(format(args)).resolves.toBeUndefined();

  expect(stdout(new RegExp(tempDir, 'g'))).toMatchSnapshot();

  await expect(gitModifiedAndUnstaged(tempDir)).resolves.toStrictEqual(
    modified,
  );

  const files = await Promise.all(
    modified.map(async (filename) => {
      const [oldFile, newFile] = await Promise.all([
        git.readBlob({
          dir: tempDir,
          fs,
          oid: originalFiles[filename],
        }),
        fs.promises.readFile(filename, 'utf8'),
      ]);

      return [
        filename,
        '-'.repeat(filename.length),
        diff(new TextDecoder().decode(oldFile.blob), newFile, {
          contextLines: 1,
          expand: false,
          omitAnnotationLines: true,
        }),
      ].join('\n');
    }),
  );

  expect(`\n${files.join('\n\n')}`).toMatchSnapshot();

  expect(process.exitCode).toBe(exitCode);
});
