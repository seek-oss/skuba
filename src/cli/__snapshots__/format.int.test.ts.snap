// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixable 1`] = `
"
skuba lints
Updating skuba...

Patch skipped: Add empty exports to Jest files for compliance with TypeScript isolated modules

Patch skipped: Update Renovate config to support private SEEK packages - owner does not map to a SEEK preset

Patch skipped: Upgrade Node.js Distroless Docker image to -debian12 variant - no Dockerfile found

Patch skipped: Add keepAliveTimeout to server listener - no listener file found

Patch skipped: Move .npmrc out of the .gitignore managed section - no .gitignore file found

Patch skipped: Move .npmrc out of the .dockerignore managed section - no .dockerignore file found

Patch skipped: Ensure the pnpm package manager version specified in package.json is used in Dockerfiles - no packageManager declaration in package.json found

Patch skipped: Remove version field from docker-compose files - no docker-compose files found

skuba update complete.

Refreshed .eslintignore. refresh-config-files
Refreshed .gitignore. refresh-config-files
Refreshed .prettierignore. refresh-config-files
Refreshed .npmrc. refresh-config-files
Refreshed .dockerignore. refresh-config-files
Processed skuba lints in <random>s.

ESLint
Processed 4 files in <random>s.

Prettier
Processed 8 files in <random>s.
Formatted 3 files:
b.md
c.json
d.js
"
`;

exports[`fixable 2`] = `
"
a/a/a.ts
--------
@@ -1,4 +1,4 @@
  // Imports out-of-order
- import path from 'path';
  import fs from 'fs';
+ import path from 'path';


b.md
----
@@ -2,3 +2,3 @@

- Trailing space 
+ Trailing space


c.json
------
@@ -2,3 +2,3 @@
    "key": "value"
-   }
+ }


d.js
----
  // eslint-disable-next-line no-console
-     console.log(process.argv);
+ console.log(process.argv);
"
`;

exports[`ok --debug 1`] = `
"
skuba lints
Updating skuba...

Patch skipped: Add empty exports to Jest files for compliance with TypeScript isolated modules

Patch skipped: Update Renovate config to support private SEEK packages - owner does not map to a SEEK preset

Patch skipped: Upgrade Node.js Distroless Docker image to -debian12 variant - no Dockerfile found

Patch skipped: Add keepAliveTimeout to server listener - no listener file found

Patch skipped: Move .npmrc out of the .gitignore managed section - no .gitignore file found

Patch skipped: Move .npmrc out of the .dockerignore managed section - no .dockerignore file found

Patch skipped: Ensure the pnpm package manager version specified in package.json is used in Dockerfiles - no packageManager declaration in package.json found

Patch skipped: Remove version field from docker-compose files - no docker-compose files found

skuba update complete.

Refreshed .eslintignore. refresh-config-files
Refreshed .gitignore. refresh-config-files
Refreshed .prettierignore. refresh-config-files
Refreshed .npmrc. refresh-config-files
Refreshed .dockerignore. refresh-config-files
Processed skuba lints in <random>s.

ESLint
Initialising ESLint...
Processing files...
Processed 2 files in <random>s.
○ a/a/a.ts
○ d.js

Prettier
Initialising Prettier...
Detected project root: <random>
Discovering files...
Discovered 11 files.
Formatting files...
.dockerignore
  parser: -
.eslintignore
  parser: -
.gitignore
  parser: -
.npmrc
  parser: -
.prettierignore
  parser: -
b.md
  parser: markdown
c.json
  parser: json
d.js
  parser: babel
package.json
  parser: json-stringify
tsconfig.json
  parser: json
a/a/a.ts
  parser: typescript
Processed 6 files in <random>s.
"
`;

exports[`ok --debug 2`] = `
"
"
`;

exports[`ok 1`] = `
"
skuba lints
Updating skuba...

Patch skipped: Add empty exports to Jest files for compliance with TypeScript isolated modules

Patch skipped: Update Renovate config to support private SEEK packages - owner does not map to a SEEK preset

Patch skipped: Upgrade Node.js Distroless Docker image to -debian12 variant - no Dockerfile found

Patch skipped: Add keepAliveTimeout to server listener - no listener file found

Patch skipped: Move .npmrc out of the .gitignore managed section - no .gitignore file found

Patch skipped: Move .npmrc out of the .dockerignore managed section - no .dockerignore file found

Patch skipped: Ensure the pnpm package manager version specified in package.json is used in Dockerfiles - no packageManager declaration in package.json found

Patch skipped: Remove version field from docker-compose files - no docker-compose files found

skuba update complete.

Refreshed .eslintignore. refresh-config-files
Refreshed .gitignore. refresh-config-files
Refreshed .prettierignore. refresh-config-files
Refreshed .npmrc. refresh-config-files
Refreshed .dockerignore. refresh-config-files
Processed skuba lints in <random>s.

ESLint
Processed 2 files in <random>s.

Prettier
Processed 6 files in <random>s.
"
`;

exports[`ok 2`] = `
"
"
`;

exports[`unfixable 1`] = `
"
skuba lints
Updating skuba...

Patch skipped: Add empty exports to Jest files for compliance with TypeScript isolated modules

Patch skipped: Update Renovate config to support private SEEK packages - owner does not map to a SEEK preset

Patch skipped: Upgrade Node.js Distroless Docker image to -debian12 variant - no Dockerfile found

Patch skipped: Add keepAliveTimeout to server listener - no listener file found

Patch skipped: Move .npmrc out of the .gitignore managed section - not ignored

Patch skipped: Move .npmrc out of the .dockerignore managed section - no .dockerignore file found

Patch skipped: Ensure the pnpm package manager version specified in package.json is used in Dockerfiles - no packageManager declaration in package.json found

Patch skipped: Remove version field from docker-compose files - no docker-compose files found

skuba update complete.

Refreshed .eslintignore. refresh-config-files
Refreshed .gitignore. refresh-config-files
Refreshed .prettierignore. refresh-config-files
Refreshed .npmrc. refresh-config-files
Processed skuba lints in <random>s.

ESLint
Processed 2 files in <random>s.

<random>/a/a/a.ts
  8:3  error  Unexpected console statement  no-console

<random>/d.js
  1:0  error  Parsing error: Unexpected token (1:0)

✖ 2 problems (2 errors, 0 warnings)


Prettier
Processed 6 files in <random>s.
Formatted 1 file:
a/a/a.ts
Flagged 1 file:
d.js SyntaxError: Unexpected token (1:1)
> 1 | }
    | ^
  2 |

ESLint, Prettier found issues that require triage.
"
`;

exports[`unfixable 2`] = `
"
a/a/a.ts
--------
@@ -1,4 +1,4 @@
  // Imports out-of-order
- import path from 'path';
  import fs from 'fs';
+ import path from 'path';

@@ -7,3 +7,3 @@

-   console.log('I forgot to remove this')
+   console.log('I forgot to remove this');
  };"
`;
